### PR TITLE
state: add change.LaneTasks helper

### PR DESCRIPTION
**Pull request by stolowski**
Implemented a helper which returns all tasks from given lanes. This will be use in #4158.

### CODE REVIEW

File: overlord/state/change.go [hunk 1]
```diff
@@ -477,6 +477,31 @@ func (c *Change) Tasks() []*Task {
 	return c.state.tasksIn(c.taskIDs)
 }
 
+// LaneTasks returns all tasks from given lanes ths state change depends on.
```
Comment (bboozzoo) [inline overlord/state/change.go]: s/ths/the/?

Comment (stolowski) [reply overlord/state/change.go]: Thanks, fixed.

File: overlord/state/change.go [hunk 2]
```diff
@@ -477,6 +477,31 @@ func (c *Change) Tasks() []*Task {
 	return c.state.tasksIn(c.taskIDs)
 }
 
+// LaneTasks returns all tasks from given lanes the state change depends on.
+// All tasks are returned if any of the lanes is 0 (the default lane).
+func (c *Change) LaneTasks(lanes []int) []*Task {
```
Comment (mvo5) [inline overlord/state/change.go]: Quick idea about the API, you could use: `LaneTasks(lanes...int) []*Task` here. This makes using the API a bit more natural, i.e. "LaneTasks(1,2)" or "LaneTasks(1)". I.e.:
```
diff --git a/overlord/state/change.go b/overlord/state/change.go
index 9246b1b6e..5b167b952 100644
--- a/overlord/state/change.go
+++ b/overlord/state/change.go
@@ -479,7 +479,7 @@ func (c *Change) Tasks() []*Task {
 
 // LaneTasks returns all tasks from given lanes the state change depends on.
 // All tasks are returned if any of the lanes is 0 (the default lane).
-func (c *Change) LaneTasks(lanes []int) []*Task {
+func (c *Change) LaneTasks(lanes ...int) []*Task {
        laneLookup := make(map[int]bool)
        for _, l := range lanes {
                if l == 0 {
diff --git a/overlord/state/change_test.go b/overlord/state/change_test.go
index b6a2aea3f..2964799b4 100644
--- a/overlord/state/change_test.go
+++ b/overlord/state/change_test.go
@@ -194,18 +194,16 @@ func (cs *changeSuite) TestLaneTasks(c *C) {
                c.Assert(tasks1, DeepEquals, tasks2)
        }
 
-       c.Assert(chg.LaneTasks([]int{}), HasLen, 0)
-
-       tasks := chg.LaneTasks([]int{0})
+       tasks := chg.LaneTasks(0)
        checkTasks(tasks, []*state.Task{t1, t2, t3, t4, t5})
 
-       tasks = chg.LaneTasks([]int{lane1})
+       tasks = chg.LaneTasks(lane1)
        checkTasks(tasks, []*state.Task{t1, t2, t4})
 
-       tasks = chg.LaneTasks([]int{lane2})
+       tasks = chg.LaneTasks(lane2)
        checkTasks(tasks, []*state.Task{t3, t4})
 
-       tasks = chg.LaneTasks([]int{lane1, lane2})
+       tasks = chg.LaneTasks(lane1, lane2)
        checkTasks(tasks, []*state.Task{t1, t2, t3, t4})
 }
 

```

File: overlord/state/change.go [hunk 3]
```diff
@@ -477,6 +477,31 @@ func (c *Change) Tasks() []*Task {
...
+		}
+		laneLookup[l] = true
+	}
+
+	c.state.reading()
+	var tasks []*Task
+	for _, tid := range c.taskIDs {
+		t := c.state.tasks[tid]
+		for _, l := range t.lanes {
+			if _, ok := laneLookup[l]; ok {
```
Comment (mvo5) [inline overlord/state/change.go]: Because this is a map of bool you could just write: `if laneLookup[l] {` here.

**Comment by stolowski** [comment]
@mvo5 Thanks for the review, nice suggestions, done.

File: overlord/state/change.go [hunk 4]
```diff
@@ -477,6 +477,31 @@ func (c *Change) Tasks() []*Task {
 	return c.state.tasksIn(c.taskIDs)
 }
 
+// LaneTasks returns all tasks from given lanes the state change depends on.
+// All tasks are returned if any of the lanes is 0 (the default lane).
+func (c *Change) LaneTasks(lanes ...int) []*Task {
+	laneLookup := make(map[int]bool)
+	for _, l := range lanes {
+		if l == 0 {
```
Comment (pedronis) [inline overlord/state/change.go]: I said this, but then I think I said it's wrong...  we don't want this. The 0 lane is not a lane where all tasks of a change are, it's the lane of all the tasks that didn't join a specific set of lanes

File: overlord/state/change_test.go [hunk 5]
```diff
@@ -144,6 +145,70 @@ func (cs *changeSuite) TestStatusExplicitlyDefined(c *C) {
...
+
+		sort.Strings(tasks1)
+		sort.Strings(tasks2)
+
+		c.Assert(tasks1, DeepEquals, tasks2)
+	}
+
+	c.Assert(chg.LaneTasks(), HasLen, 0)
+
+	tasks := chg.LaneTasks(0)
```
Comment (pedronis) [inline overlord/state/change_test.go]: we need a different test/behavior for this as explained

File: overlord/state/change.go (hunk 4 above)
Comment (stolowski) [reply overlord/state/change.go]: Ok, should be fixed now.

**Comment by stolowski** [comment]
Ready for re-review.

File: overlord/state/change.go [hunk 6]
```diff
@@ -477,6 +477,32 @@ func (c *Change) Tasks() []*Task {
 	return c.state.tasksIn(c.taskIDs)
 }
 
+// LaneTasks returns all tasks from given lanes the state change depends on.
+// All tasks are returned if any of the lanes is 0 (the default lane).
```
Comment (pedronis) [inline overlord/state/change.go]: the doc comment needs updating to remove the 2nd sentence.

Comment (stolowski) [reply overlord/state/change.go]: Done

**Pull request merged**